docs/src/main/tut/datatypes/validated.md

```diff
@@ -12,7 +12,351 @@ Response comes back saying your username can't have dashes in it, so you make so
+    if (firstName.matches("^[a-zA-Z]+$")) Right(firstName) else Left(FirstNameHasSpecialCharacters)
+
+  private def validateLastName(lastName: String): Either[DomainValidation, String] =
+    if (lastName.matches("^[a-zA-Z]+$")) Right(lastName) else Left(LastNameHasSpecialCharacters)
+
+  private def validateAge(age: Int): Either[DomainValidation, Int] =
+    if (age >= 18 && age <= 75) Right(age) else Left(AgeIsInvalid)
+
+  def validateForm(username: String, password: String, firstName: String, lastName: String, age: Int): Either[DomainValidation, RegistrationData] = {
+
+    for {
+      validatedUserName <- validateUserName(username)
+      validatedPassword <- validatePassword(password)
+      validatedFirstName <- validateFirstName(firstName)
+      validatedLastName <- validateLastName(lastName)
+      validatedAge <- validateAge(age)
+    }
+      yield RegistrationData(validatedUserName, validatedPassword, validatedFirstName, validatedLastName, validatedAge)
+  }
+
+}
+
+object FormValidator extends FormValidator
+
+```
+
+The logic of the validation process is as follows: **check every individual field based on the established rules for each one of them. If the validation is successful, then return the field wrapped in a `Right` instance; If not, then return a `DomainValidation` with the respective message, wrapped in a `Left` instance**.
+
+Our service has the `validateForm` method for checking all the fields and, if the process succeeds it will create an instance of `RegistrationData`, right?
+
+Well, yes, but the error reporting part will have the downside of showing only the first error.
+
+Let's look in detail this part:
+
+```tut:silent:fail
+for {
+  validatedUserName <- validateUserName(username)
+  validatedPassword <- validatePassword(password)
+  validatedFirstName <- validateFirstName(firstName)
+  validatedLastName <- validateLastName(lastName)
+  validatedAge <- validateAge(age)
+}
+  yield RegistrationData(validatedUserName, validatedPassword, validatedFirstName, validatedLastName, validatedAge)
+```
+
+A for-comprehension is _fail-fast_. If some of the evaluations in the `for` block fails for some reason, the `yield` statement will not complete. In our case, if that happens we won't be getting the accumulated list of errors.
+
+If we run our code:
+
+```tut:book
+FormValidator.validateForm(
+  username = "fakeUs3rname", 
+  password = "password", 
+  firstName = "John", 
+  lastName = "Doe",
+  age = 15
+)
+```
+
+We should have gotten another `DomainValidation` object denoting the invalid age.
+
+### An iteration with `Validated`
+
+Time to do some refactoring! We're going to try a `Validated` approach:
+
+```tut:silent
+import cats.data._
+import cats.data.Validated._
+import cats.implicits._
+
+def validateUserName(userName: String): Validated[DomainValidation, String] = {
+  if (userName.matches("^[a-zA-Z0-9]+$")) Valid(userName) else Invalid(UsernameHasSpecialCharacters)
+}
```

Comment:
Maybe reuse the `Either` methods here and call `toValidated` ?

Reply:
Good suggestion. I'll take a look at it!